specification.md

```diff
@@ -1563,3 +1571,43 @@ to send over the connection. When using a single connection to transmit
 larger messages, TCP buffers can be more efficiently utilized.
 Additionally, it becomes feasible to tune buffer sizes in order to
 optimize the throughput of connections to particular services.
+
+
+
+### Appendix D - Environment Variable Expansion
+
+Environment variable expansion implies that certain strings, such as values in
+the [environment dictionary](#jobspecification-setenvironment) are subject to
+environment variable substitution rules. That is, sequences of the form
+`${VAR_NAME}` are replaced with the value of the environment variable `VAR_NAME`
+as defined on the machine where the job runs. For multi-process jobs (e.g. MPI
+jobs), the *machine* is defined as being the rank 0 node.
+
+Environment variable expansion in the environment dictionary is essential in
+being able to express common patterns, such as extending, for example, the
+`PATH` or `LD_LIBRARY_PATH` environment variables trough code such as:
+
+```python
+jobspec.environment = {'PATH': '/home/${USER}/app:${PATH}'}
+```
+
+In a `fork()/execve()` based local implementation of a
+[JobExecutor](#jobexeuctor) adapter, the adapter would take care of parsing
+values in the `jobspec.environment` dictionary and substituting relevant
+variables from the calling process' `environ`. In an adapter in which submission
+is implemented through submit scripts that are passed to a LRM, an
+implementation can simply leave the values in `jobspec.environment` as is (after
+properly quoting expressions that are not environment variable dereferences in
+this specification). A remote layer (i.e. Layer 1) implemented on top of
+existing Layer 0 adapters would rely exclusively on the Layer 0 adapter to
+implement environment variable expansion, whereas a pilot job system (Layer 2)
+would need to implement this feature in the pilot job.
+
+A question of relevance is whether adding support for environment variable
+expansion in strings other than values in `jobspec.environment` places an an
+unreasonable additional burden on the implementation. The answer appears to be
+negative. If an adapter implementation performs expansion on values in
+`jobspec.environment`, it must already have the tools that perform the parsing
+and token substitution on arbitrary strings. These tools can equally be applied
+to strings in `jobspec.arguments`. A similar argument can be made about Layer 0
+adapters that are based on submit scripts.
```

Comment:
Given our recent discussion of `pre_launch` and `pre_exec`, we could consider to restrict the `environment` setting to trivial var names, and move ano more complex env settings (which requires recursive and ordered expansion) to the `pre_*` attributes.

One problem with env expansion in the `environment` attribute is that we would need ordering.  A dict version (unordered) would be ambiguous:
```py
j.environment = {'FOO' : FOO_'$BAR', 
                           'BAR': 'BAR_$FOO'}
```
Without ordering, the results could be
```sh
FOO = FOO_
BAR = BAR_FOO_
```
or
```
FOO = FOO_BAR_
BAR = BAR_
```

I think we discussed this on the call, so just recording that point here...